[RA2 Ch3] CPU Management (#2147)
cntt-n/CNTT: doc/ref_arch/kubernetes/chapters/chapter03.md

> This chapter should describe considerations about hardware acceleration, like
device management.

<a name="3.2.1.2"></a>
#### 3.2.1.2 CPU Management

> CPU management has policies to determine placement preferences to use for workloads that are sensitive to cache affinity or latency, so must not be moved by OS scheduler or throttled by kubelet. Additionally, some workloads are sensitive to differences between physical cores and SMT, while others (like DPDK workloads) are designed to run on isolated CPUs (like Linux isolcpus kernel parameter specifying cores isolated from general SMB balancing and scheduler algorighms).

Reviewer: It is confusing as to what should "not be moved"?
For DPDK, in addition to isolcpus should cset be mentioned also?
Minor typo: SMB instead of SMP
```suggestion
CPU management has policies to determine placement preferences to use for workloads that are sensitive to cache affinity or latency, and so the workloads must not be moved by OS scheduler or throttled by kubelet. Additionally, some workloads are sensitive to differences between physical cores and SMT, while others (like DPDK workloads) are designed to run on isolated CPUs (like Linux isolcpus kernel parameter specifying cores isolated from general SMP balancing and scheduler algorithms).
```

Author: > 
> 
> It is confusing as to what should "not be moved"?
> Minor typo: SMB instead of SMP

Thanks for clarifying and fixing.

> For DPDK, in addition to isolcpus should cset be mentioned also?

Added "with cpuset-based selection of CPUs".

> CPU management has policies to determine placement preferences to use for workloads that are sensitive to cache affinity or latency, and so the workloads must not be moved by OS scheduler or throttled by kubelet. Additionally, some workloads are sensitive to differences between physical cores and SMT, while others (like DPDK-based workloads) are designed to run on isolated CPUs (like on Linux with cpuset-based selection of CPUs and isolcpus kernel parameter specifying cores isolated from general SMP balancing and scheduler algorithms).

Reviewer 2: ```suggestion
CPU management has policies to determine placement preferences to use for workloads that are sensitive to cache affinity or latency, and so the workloads must not be moved by OS scheduler or throttled by kubelet. Additionally, some workloads are sensitive to differences between physical cores and SMT, while others (like DPDK-based workloads) are designed to run on isolated CPUs (like on Linux with cpuset-based selection of CPUs and isolcpus kernel parameter specifying cores isolated from general SMP balancing and scheduler algorithms) or required to run on the same NUMA domain as some of their other resources.
```

Author: @CsatariGergely : This is of course correct text, just I'm not sure it fits here well. In this section I was focusing only on CPU mgmt, and other requirements described in their sections. (NUMA) Topology in #2150 already has similar text, and more we could add as needed to clarify. How apps use all of those features together is in #2154 . Please check again and let me know your recommendation.

Reviewer 2: I would keep it for the shake of completeness in this section, but if you really do not like it I can also accept that.
I resolve the conversation and let you to decide ... :)

Reviewer 3: @petorre can you let me know if you want to accept the change, or continue with your text?

> Kubernetes [CPU Manager](https://kubernetes.io/docs/tasks/administer-cluster/cpu-management-policies/) works with Topology Manager. Special care needs to be taken of:

Reviewer 3: Should we explain what `Topology Manager` is, or provide a link?

Author: > 
> 
> Should we explain what `Topology Manager` is, or provide a link?

It is explained and linked in #2150 for 3.2.1.4 Hardware Topology Management .

> •	Supporting isolated CPUs: Using kubelet [Reserved CPUs](https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/#explicitly-reserved-cpu-list) and Linux isolcpus allows configuration where all allocatable CPUs are isolcpus, and pod scheduling on such nodes can be influenced with taints, tolerations and node affinity.

Reviewer: @petorre I am a bit confused. The starting topic is "Supporting isolated CPUs" but the reference is to "Reserved CPUs" and then the statement, "all allocatable CPUs are isolcpus." As you know isolated CPUs are not schedulable and thus can't be influenced as they need to be manually managed using taskset. What am I getting wrong?

Author: > @petorre I am a bit confused. The starting topic is "Supporting isolated CPUs" but the reference is to "Reserved CPUs" and then the statement, "all allocatable CPUs are isolcpus." As you know isolated CPUs are not schedulable and thus can't be influenced as they need to be manually managed using taskset. What am I getting wrong?

@pgoyal01 Please see reworded text split into two sentences.

> •	Differentiating between physical cores and SMT: When requesting even number of CPU cores for pods, scheduling can be influenced with taints, tolerations and node affinity.

Reviewer 4: ```suggestion
•	Differentiating between physical cores and SMT: When requesting even number of CPU cores for pods, scheduling can be influenced with taints, tolerations, and node affinity.
```

> •	Differentiating between physical cores and SMT: When requesting even number of CPU cores for pods, scheduling can be influenced with taints, tolerations and node affinity.

Reviewer 4: ```suggestion
•	Differentiating between physical cores and SMT: When requesting even number of CPU cores for pods, scheduling can be influenced with taints, tolerations, and node affinity.
```

Reviewer 3: @petorre do you want to accept/apply this change?

Author: > @petorre do you want to accept/apply this change?

@tomkivlin : Not sure which other change made it that way but the current text already has that "," included. Thanks

Reviewer 3: OK missed that, sorry.
